csharp/src/Ice/ConnectionI.cs

```diff
@@ -607,15 +607,23 @@ public override bool startAsync(int operation, Ice.Internal.AsyncCallback comple
                             observerStartWrite(_writeStream.getBuffer());
                         }
 
-                        bool completed;
-                        if (_transceiver.startWrite(_writeStream.getBuffer(), completedCallback, this, out completed))
+                        bool messageWritten = false;
+                        bool completedSynchronously =
+                            _transceiver.startWrite(
+                                _writeStream.getBuffer(),
+                                completedCallback,
+                                this,
+                                out messageWritten);
```

Comment:
It's not clear to me why the C# startWrite has such a complicated signature compared to the C++ startWrite. In C++, we don't have this "completedSynchronously" return value. 

In any case, assuming this complicated C# API is warranted, we should:
 - rename the last parameter of Transceiver.startWrite. Having return value = completedSynchronously and out parameter = completed is too confusing.
 - consider returning a tuple (does this work with the old C# version we use on 3.7?)
 - name the "messageWritten" parameter => messageFullyWritten
 - add a real doc-comment on Transceiver.startWrite

Reply:
Yes, this wait until a follow-up PR.